### DISCUSSION TITLE
Trying to return a PDF through the Serializer

### DISCUSSION THREAD

**Discussion by TristanPouliquen** [symfony/symfony#43234] [Q&A] [answered]
Hello,

I'm trying to see if it is easily feasible to add the PDF format as an alternative for some endpoints of my API (resources through API Platform).

As API Platform relies on the Symfony Serializer, I tried to define a custom Normalizer & Encoder that would receive my entities, decide on which template & which context to pass to Twig to generate the HTML & return a `BinaryFileResponse`.

The PDFs are successfully created, saved locally.

At this point, I thought that all would work seamlessly after reading the (`SerializeListener`)[https://github.com/api-platform/core/blob/db5275885a3052b52a60dedcf530d06199a16f01/src/Core/EventListener/SerializeListener.php#L121] and (`RespondListener`)[https://github.com/api-platform/core/blob/db5275885a3052b52a60dedcf530d06199a16f01/src/Core/EventListener/RespondListener.php#L69].

It didn't work.

By diving into it, I realized that, when I do a `gettype($event->getControllerResult())` in the `SerializeListener` just after having added the return from the Serializer, I get a `string`.

It seems that somewhere, my `BinaryFileResponse` is stringified, which loses all interest for me. (maybe because of the typing on the higher levels of the Serializer that demand a string return from encoders? I did not see it at first from PHPStorm as it is only set in PHPDoc at the EncoderInterface level).

So, does someone know of a way to make all this work magically with the Symfony Serializer? Or should I fall back to uglier custom controllers for these endpoints?

I found much more beautiful to be able to do `GET /api/entities/{id}.pdf` to get my PDF version of the entity than have to define an extra route and all. And doing it via the Serializer would have allowed me to abstract a huge bit of our PDF generation system.

**Comment by stof** [comment, marked as answer]
> It seems that somewhere, my `BinaryFileResponse` is stringified, which loses all interest for me. (maybe because of the typing on the higher levels of the Serializer that demand a string return from encoders? I did not see it at first from PHPStorm as it is only set in PHPDoc at the EncoderInterface level).

Encoders are indeed meant to return strings. Symfony 6 will have a native return type there (the `EncoderInterface` was added in Symfony at a time return types were not available in PHP)

**Reply by TristanPouliquen** [reply]
So no way to encode in anything else than a string through the Symfony Serializer ?
